### PR TITLE
views: file listing based on permissions

### LINKED ISSUE #423: migration: record file download
- [ ] Documents/Records-UI integration (access control, list files).
- [x] Template: Display files on record page.
- [x] Access control: Closed access files (owner, admin, shared record link), Embargoed (owner, admin, shared record link, after date).

### PR DESCRIPTION
**Pull request by SamiHiltunen**
* Lists files in record view only if the viewer has a permission to
  view them. (closes #423)

Signed-off-by: Sami Hiltunen <sami.mikael.hiltunen@cern.ch>

### CODE REVIEW

File: zenodo/modules/frontpage/templates/zenodo_frontpage/index.html [hunk 1]
```diff
@@ -44,7 +44,7 @@
         {%- set publication_date = r.publication_date|from_isodate -%}
         {%- set creation_date = r._created|from_isodatetime -%}
         {%- set embargo_date = r.embargo_date|from_isodate -%}
-        {%- set object_type = r.resource_type|objecttype %}
+        {%- set object_type = r.resource_type.to_dict()|objecttype %}
```
Comment (lnielsen) [inline zenodo/modules/frontpage/templates/zenodo_frontpage/index.html]: Do you know why this one is needed? It's probably better to fix in the ``objecttype`` filter or on ``ObjectType`` model.

Comment (SamiHiltunen) [inline zenodo/modules/frontpage/templates/zenodo_frontpage/index.html]: Elasticsearch-dsl uses an AttrDict which does have a get method. I added this here for now but the proper fix should most likely go into elasticsearch-dsl
https://github.com/elastic/elasticsearch-dsl-py/blob/master/elasticsearch_dsl/utils.py#L76

File: zenodo/modules/records/models.py [hunk 2]
```diff
@@ -68,6 +88,9 @@ class AccessRight(object):
     @staticmethod
     def is_embargoed(embargo_date):
         """Test if date is still under embargo."""
+        embargo_date = embargo_date \
+            if not isinstance(embargo_date, string_types) \
+            else parser.parse(embargo_date).date()
```
Comment (lnielsen) [inline zenodo/modules/records/models.py]: For date parsing it's better to use arrow library for consistent among our modules (under the hood it uses dateutil):
```python
import arrow
arrow.get(value).date()
```

File: zenodo/modules/frontpage/templates/zenodo_frontpage/index.html (hunk 1 above)
Comment (lnielsen) [inline zenodo/modules/frontpage/templates/zenodo_frontpage/index.html]: I agree that the fix is for es dsl - can you open an issue, but in the mean time move the logic to ``ObjectType`` model (e.g. the ``to_dict()`` i don't think will work if the record suddenly comes from the db).

File: zenodo/modules/records/models.py [hunk 3]
```diff
@@ -68,7 +88,7 @@ class AccessRight(object):
     @staticmethod
     def is_embargoed(embargo_date):
         """Test if date is still under embargo."""
-        return embargo_date > datetime.utcnow().date()
+        return arrow.get(embargo_date).date() > datetime.utcnow().date()
```
Comment (lnielsen) [inline zenodo/modules/records/models.py]: Does this work with both ``datetime`` and ``string`` objects?

File: zenodo/modules/records/permissions.py [hunk 4]
```diff
@@ -45,18 +45,30 @@ def __init__(self, bucket, action='objects-read'):
...
+            == AccessRight.OPEN:
+        return True
 
-        return False
+    user_id = int(user.get_id()) if user.is_authenticated else None
+
+    if user_id in record.get('owners', []):
+        return True
+
+    identity = Identity(user)
```
Comment (lnielsen) [inline zenodo/modules/records/permissions.py]: Is the ``UserNeed`` not automatically provided by Flask-Security on login? - https://github.com/mattupstate/flask-security/blob/develop/flask_security/core.py#L213

File: zenodo/modules/records/models.py (hunk 3 above)
Comment (SamiHiltunen) [inline zenodo/modules/records/models.py]: This works with both datetime and string. I added a test to verify that.

File: zenodo/modules/records/permissions.py (hunk 4 above)
Comment (SamiHiltunen) [inline zenodo/modules/records/permissions.py]: I loaded the identity so has_access could be called to verify user's access to a record without having to have the user currently logged in. I removed it now, should be easy to add it later if this is actually needed.

File: zenodo/modules/frontpage/templates/zenodo_frontpage/index.html (hunk 1 above)
Comment (SamiHiltunen) [inline zenodo/modules/frontpage/templates/zenodo_frontpage/index.html]: I moved the check to ObjectType. Seems they removed the get method on purpose earlier. https://github.com/elastic/elasticsearch-dsl-py/commit/62ce0fc3684709e1370e00554b706eda53220075

File: tests/unit/records/test_models.py [hunk 5]
```diff
@@ -111,6 +111,13 @@ def test_access_right_embargo():
         AccessRight.EMBARGOED, embargo_date=today+timedelta(days=1)) \
         == 'embargoed'
 
+    # Should work with strings as well
```
Comment (lnielsen) [inline tests/unit/records/test_models.py]: 👍

File: zenodo/modules/records/models.py [hunk 6]
```diff
@@ -165,6 +196,8 @@ def get_by_dict(cls, value):
         if not value:
             return None
         if 'subtype' in value:
+            if isinstance(value, AttrDict):
```
Comment (lnielsen) [inline zenodo/modules/records/models.py]: 👍

**Comment by lnielsen** [comment]
LGTM - just needs a rebase.

**Comment by SamiHiltunen** [comment]
@lnielsen rebased

**Pull request merged**